libraries/libstratosphere/source/settings/impl/settings_key_value_store.cpp

```diff
@@ -0,0 +1,1629 @@
+                    R_TRY(GetSystemData<SystemDataTag::PfCfg>(std::addressof(data), ncm::SystemDataId::PlatformConfigHoag));
+                    break;
+                case SplHardwareType_Calcio:
+                    R_TRY(GetSystemData<SystemDataTag::PfCfg>(std::addressof(data), ncm::SystemDataId::PlatformConfigCalcio));
+                    break;
+                case SplHardwareType_Aula:
+                    R_TRY(GetSystemData<SystemDataTag::PfCfg>(std::addressof(data), ncm::SystemDataId::PlatformConfigAula));
+                    break;
+                AMS_UNREACHABLE_DEFAULT_CASE();
+            }
+
+            /* Ensure data is not null. */
+            AMS_ASSERT(data != nullptr);
+
+            /* Load the key value store map. */
+            return LoadKeyValueStoreMapDefault<SystemData>(out, *data);
+        }
+
+        template<typename T>
+        Result LoadKeyValueStoreMapCurrent(Map *out, T &data) {
+            /* Check pre-conditions. */
+            AMS_ASSERT(out != nullptr);
+
+            /* Open the data for reading. */
+            R_TRY(data.OpenToRead());
+
+            /* Load the map entries. */
+            R_TRY(LoadKeyValueStoreMapEntries(out, data, [](Map &map, const MapKey &key, u8 type, const void *value_buffer, u32 value_size) -> Result {
+                AMS_UNUSED(type);
+                /* Find the key in the map. */
+                if (auto it = map.find(key); it != map.end()) {
+                    MapValue &map_value = it->second;
+                    size_t current_value_size = value_size;
+                    void *current_value_buffer = nullptr;
+
+                    if (value_size > 0) {
+                        /* Ensure there is sufficient memory for the value. */
+                        R_UNLESS(GetHeapAllocatableSize() >= current_value_size, ResultSettingsItemValueAllocationFailed());
+                    
+                        /* Allocate the value buffer. */
+                        current_value_buffer = AllocateFromHeap(current_value_size);
+                        AMS_ASSERT(current_value_buffer != nullptr);
+
+                        /* Copy the value to the current value buffer. */
+                        std::memcpy(current_value_buffer, value_buffer, current_value_size);
+                    }
+
+                    /* Replace the current value buffer with a new one. */
+                    std::swap(map_value.current_value_size, current_value_size);
+                    std::swap(map_value.current_value, current_value_buffer);
+
+                    /* Free the old buffer if it is no longer in use. */
+                    if (current_value_buffer != nullptr && current_value_buffer != map_value.default_value) {
+                        FreeToHeap(current_value_buffer, current_value_size);
+                    }
+                }
+
+                return ResultSuccess();
+            }));
+
+            /* Close the data. */
+            data.Close();
```

Comment:
You have a correctness issue. This needs to be make ON_SCOPE_EXIT { data.Close(); }; after data.OpenToRead();